### DISCUSSION TITLE
weighted search on multiple keys

### DISCUSSION THREAD

**Discussion by drydenwilliams** [krisk/Fuse#590] [Q&A] [answered]
I wondered if it's possible using Fuse to search multiple values and keys? 

I would like to sort this list of jobs to find the closest one for a search. 

```
const jobs = [
  {
    title: "Web developer",
    people: 0,
    physical: 0,
    mental: 10,
    location: 0,
    money: 8,
    value: 0
  },
  {
    title: "Web developer in a studio",
    people: 8,
    physical: 0,
    mental: 5,
    location: 0,
    money: 10,
    value: 0
  },
  {
    title: "Carpenter",
    people: 2,
    physical: 10,
    mental: 8,
    location: 0,
    money: 7,
    value: 4
  },
  {
    title: "Teacher",
    people: 10,
    physical: 3,
    mental: 8,
    location: 8,
    money: 4,
    value: 7
  },
  {
    title: "Journalist",
    people: 8,
    physical: 1,
    mental: 7,
    location: 6,
    money: 5,
    value: 4
  }
];
```

And if the user was searching for a job with: 

```
{
    people: 8,
    physical: 1,
    mental: 7,
    location: 6,
    money: 5,
    value: 4
}
```

It would show the "Journalist" object as it's the closest to all weighted keys. I'm just not quite sure how to do this in general and wondered If I could use Fuse for this?

**Comment by krisk** [comment]
You _could_ use [extended search](https://fusejs.io/examples.html#extended-search). Perhaps you may also want to pair it with [logical query operations](https://fusejs.io/api/query.html)

**Comment by krisk** [comment, marked as answer]
Looking at this again, my earlier suggestion wasn't great for this use case — this is really a numeric distance/nearest-neighbor problem, not a text search one. Fuse operates on string matching, so it's not the right tool for comparing numeric attribute vectors.

For this kind of thing, you'd want to compute something like Euclidean distance between the query object and each job, then sort by that. A simple approach:

```js
const query = { people: 8, physical: 1, mental: 7, location: 6, money: 5, value: 4 }
const keys = Object.keys(query)

const sorted = jobs
  .map(job => ({
    item: job,
    distance: Math.sqrt(keys.reduce((sum, k) => sum + (job[k] - query[k]) ** 2, 0))
  }))
  .sort((a, b) => a.distance - b.distance)
```
